### ISSUE TITLE
[Elastic Agent] Support for input templates

### ISSUE THREAD

**Issue by ruflin** [elastic/beats#24054]
The configuration of the Elastic Agent is a list of inputs and not modules anymore like *beats had. Instead, modules are moved to integrations managed in Kibana. The Elastic Agent itself has only knowledge of inputs. This make the standalone configuration more verbose and potentially more cumbersome to setup.

To solve this issue, this is a proposal to support "input templates" in Elastic Agent. An input template is a template file which can contain one or many inputs and referenced in the Elastic Agent and variables can be passed in.

## Template file

Each template file contain a list of inputs. The most basic template is just a normal input config. For example `templates.d/nginx-logs.yml` could look as following:

```
- type: logfile
  paths: /var/log/nginx/*.log
```

The above does not contain any variables. But as it is a template, also variables should be supported:

```
- type: logfile
  paths: {{path|/var/log/nginx/*.log}}
```

NOTE: This is not a proposal on how the final config / variables should look like, it is only used as an example.

These templates could the be used in the input configs:

```
inputs:
 - type: template
   id: nginx-logs # Template file reference
   vars:
     path: /foo/bar/nginx/*.log
```

The same template can be used multiple times with different variables. Each template can also contain a list of inputs and conditions in a template can be used that only a section of the template is run if certain variables are set. This becomes useful if the `nginx` temlate contains configs for metrics and logs, but only logs should be run.

## Default templates

In the case of system metrics collection and logs, the default template for the collection should already be shipped with the Elastic Agent to make the getting started in standalone easier. The config then could look like the following:

```
- type: template
  id: system
```

This would collect the default system and metrics logs. An extension of this is that templates could also be pulled from the registry. Also users can build their own templates and put them into the template directory.

This is related to https://github.com/elastic/beats/issues/23979

**Comment by ph** [comment]
@urso Would like your opinion on this.

**Comment by urso** [comment]
In general +1 on a template feature. It would somewhat resemble lightweight modules, in a way that a template is just another input type (-> allow a template to instantiate other templates).

The exact format still needs some thought it seems.

Will we make templates available to package authors as inputs, or will templates replace some functionality in todays package spec?

**Comment by ruflin** [comment]
> Will we make templates available to package authors as inputs, or will templates replace some functionality in todays package spec?

Interesting idea. If templates become and assets it should also be possible to use it in a package? 🤔  So far my thinking was that during package build time, a template for the full package (or parts of it) is generated and added to the package. At least for our integrations and this could be a feature of `elastic-package`, that it takes all input configs and templates and puts them together. Also nice for validation / testing.

**Comment by ph** [comment]
@ruflin When "template.d/" be populated, will this be done as part of the build process or tooling to download it.

**Comment by mukeshelastic** [comment]
@ruflin 
Are we able to add an example for demostrating UX for specifying template and variables for metrics collection in agent config, I imagine it will look similar to kubernetes [config map](https://github.com/elastic/beats/blob/master/deploy/kubernetes/elastic-agent-standalone-kubernetes.yml#L89) as well as how the integration author will create the template for metrics collection.

**Comment by ruflin** [comment]
@mukeshelastic Are you looking for something more then the examples in the descriptions?

This issue by design is currently completely decoupled from packages even though packages could contain templates.

**Comment by blakerouse** [comment]
I think this work comes into 2 parts:

#### Part 1 -  Support for templates in Elastic Agent inside a top-level `templates.d` directory.

This is the core work on how Elastic Agent will perform the work required to actually support templates. This should be possible building on top of the variables used by dynamic inputs. I think the flow would go something like this.

1. Reads the configuration file
2. Splits each input into its own parsing block (already exists because of dynamic inputs) https://github.com/elastic/beats/blob/74b6e5e4a96d06b08549878a467151af033c7c2b/x-pack/elastic-agent/pkg/agent/application/emitter.go#L203
3. Checks if the input type is a template (**new**)
4. Loads the template as is without parsing any variables (**new**)
    * possible that this injects multiple inputs, so that would need to be handled
5. Clones the `transpiler.Vars` (**new**)
6. Adding a scoped `template.*` variable that comes from the `vars` section when defined by the user in the `elastic-agent.yml` configuration. (**new**)
7. Runs `Apply` with the `transpiler.Vars` created in step 6.  https://github.com/elastic/beats/blob/74b6e5e4a96d06b08549878a467151af033c7c2b/x-pack/elastic-agent/pkg/agent/application/emitter.go#L208

This would result in these 2 inputs creating this output.

`elastic-agent.yml`
```
inputs:
  - type: template
    id: my-template
    vars:
       paths:
          syslog: /var/log/custom-path
```

`templates.d/my-template.yml`
```
inputs:
  - type: logfile
    path: ${template.paths.syslog|'/var/log/syslog'}
```

`rendered config`
```
inputs:
  - type: logfile
   path: /var/log/custom-path
```

#### Part 2 -  Support for templates being shipped by default from Elastic Agent.

Because of self-upgrading of Elastic Agent the shipped templates need to be scoped to the version of the installed Elastic Agent. So all shipped templates would be built into the Elastic Agent binary at `data/elastic-agent-${ver}/templates.d`. This would need to be done at build time, pulling the templates from the package registry. We would not want to commit these templates into the beats repo.

I think the top-level `templates.d` directory would override the templates that are shipped with Elastic Agent in `data/elastic-agent-${ver}/templates.d`.

**Comment by mukeshelastic** [comment]
> @mukeshelastic Are you looking for something more then the examples in the descriptions?
> 
> This issue by design is currently completely decoupled from packages even though packages could contain templates.

@ruflin Right, I was looking for an example template definition for metrics collection.  I understand that agent wouldn't care about the contents of the template so that's why we don't need the diversity of examples in this issue. So who creates templates, what do they look like? I see your related comment above "during package build time, a template for the full package (or parts of it) is generated and added to the package." So this is generated automatically for each package? And if it does, how does the user specify in the agent config the type of datasets to gather and any other user variables. We can create another issue for the template definition and have that discussion there and track it separately.

**Comment by ruflin** [comment]
@mukeshelastic @blakerouse Lets focus on Phase 1 defined by Blake for now. @mukeshelastic For template generation of packages, should we start a discussion in the "integrations" repo?

@blakerouse @urso The part I'm not sure yet: Is it a good idea to have an input `type: template`. Or should we have separate configs? Should it be a provider? It seems the config proposed would work but will it give a good user experience?

**Comment by blakerouse** [comment]
@ruflin I like the template in the `inputs` like you have in the description. I don't like the idea of having a separate block to define inputs and then templates, especially since template only contain inputs anyway.

I would change the definition some in the `elastic-agent.yml` configuration. I would change it to:

```
inputs:
  - type: template/{name}
     ...{vars}...
```

Being that inside a template all the inputs must have a `type` defined, the type from the outside of the template cannot be overwritten. That would mean that `type` is the only key that is required on an input definition, which makes all the other keys the variables that can be substituted into the template.

Let me break it down into an example:

`elastic-agent.yml`
```
inputs:
  - type: template/system
    paths:
      syslog: /var/log/my-custom-syslog-path
```

`templates.d/system.yml`
```
inputs:
  - type: logfile
    path: ${template.paths.syslog|'/var/log/syslog'}
```

`rendered config`
```
inputs:
  - type: logfile
    path: /var/log/my-custom-syslog/path
```

This basically removes the `vars` key being defined on the template input definition. I think it makes it simpler to use and easier to read and understand. The `template/{name}` allows the overrides to work well as for templates. So `{name}` is always converted to `{name}.yaml` and looked up first in `elastic-agent-{ver}/data/templates.d` then can be override with top-level `templates.d`.

**Comment by ruflin** [comment]
I really like the idea `- type: template/{name}`. The reason I kind of like `vars` in addition is that it makes it clear, this is about variable definitions. The global part "could" be used in a different way that it could apply this section to all inputs, like adding a processor. What is the reason you prefix all variables by `template.`?

**Comment by blakerouse** [comment]
I think `vars` is overly explicit, and if you forget to add it, then any settings set at the top-level would just be ignored.

The composable inputs requires that every variable be prefixed. In this case I use `template`. That ensures that no other provider will use that name. Without the prefix a variable could override another providers. Example would be if you had a variable named `kubernetes`. It would override the `kubernetes` provider. Place the `template.*` before ensures no collision will ever happen.

**Comment by ruflin** [comment]
Dropping vars means we need to make sure we never introduce a different config option on that level or at least, it could mean a breaking change for some users. But that should be fine.

Prefix makes sense, it also means all providers can be used inside the templates which is nice. I forsee that there will be perhaps a template `nginx_k8s` and `nginx_host` and the user can select which one he wants.

**Comment by ph** [comment]
> Prefix makes sense, it also means all providers can be used inside the templates which is nice. I forsee that there will be perhaps a template nginx_k8s and nginx_host and the user can select which one he wants.

It could be different templates or a single template with conditionals.

**Comment by urso** [comment]
tl'dr; from an UX perspective I would assume that inputs and templates are configured in a similar manner.
 

> The part I'm not sure yet: Is it a good idea to have an input type: template. Or should we have separate configs? Should it be a provider? It seems the config proposed would work but will it give a good user experience?

I would prefer the UX of templates to be similar to inputs. No special configuration namespace or provider. Creating a template is an alternative way for an input author to define a new (more abstract) input type.

Same experience to inputs also implies that we should not have the `vars`, but allow the settings to be defined at the top-level. I would like templates to be configured how @blakerouse has [proposed](https://github.com/elastic/beats/issues/24054#issuecomment-786138770).

> Dropping vars means we need to make sure we never introduce a different config option on that level or at least, it could mean a breaking change for some users. But that should be fine.

I'd prefer to drop `vars`, in order to keep the template UX more consistent with the input user experience. We do have reserved settings that are special to the agent. E.g. `type` or `streams`. When it comes to reserved settings, the same issue with potential breaking changes when different configurations are introduced would exist for inputs and templates alike. Besides 'reserved' settings, the naming of settings is up to the input/template author, although we should make an attempt to keep settings consistent across inputs.

> The composable inputs requires that every variable be prefixed. In this case I use template. That ensures that no other provider will use that name. Without the prefix a variable could override another providers. Example would be if you had a variable named kubernetes. It would override the kubernetes provider. Place the template.* before ensures no collision will ever happen.

+100

> I forsee that there will be perhaps a template nginx_k8s and nginx_host and the user can select which one he wants.

I guess we have to talk about how we want to organize template alternatives in the future. But I think I would prefer to have k8s/host/X to be in a single template. So the fleet admin can enable nginx, independent of the provider someone runs locally (operator is moving from VM setup to k8s setup?).

**Comment by blakerouse** [comment]
@ph That is a good point, I never thought of that as an option. I would say to start we should not support that, but maybe in the future it would be even more powerful if we did. But for now if a template includes an `type: template/` it would cause Elastic Agent to error.

**Comment by ph** [comment]
@blakerouse Yes, it might be powerful, I'll refer to @exekias and @ChrsMark if they see a user case for it. Also until we support we could probably fake it with some tooling that generate the integration.

**Comment by ph** [comment]
@faec There is a lot to unpack in the issue, maybe we should have a kickoff?

**Comment by exekias** [comment]
> I guess we have to talk about how we want to organize template alternatives in the future. But I think I would prefer to have k8s/host/X to be in a single template. So the fleet admin can enable nginx, independent of the provider someone runs locally (operator is moving from VM setup to k8s setup?).

+1, the way I'm thinking about this is: Templates allow us to encapsulate complex config, and expose them through a small set of parameters. Ideally, no matter the scenario, you could use the same template just by adjusting these parameters, @ChrsMark outlined how this could look https://github.com/elastic/beats/issues/23876#issuecomment-794074001, note for example how the input type can be changed from `log` to `container` if needed.

**Comment by ruflin** [comment]
Agree, it would be great if we can keep all of it in a single template and I think there are multiple options to achieve this like the one outlined by @ChrsMark I suggest we go down this path for now until it breaks.

**Comment by urso** [comment]
Assuming input templates will be used for auto-discovery as well, what are potential security implications. e.g. when configuring the logs input, is it possible to splice in my own path by adding patterns like `../../../` or `\n  - ...` to template variables?

**Comment by ph** [comment]
@urso I believe you are referring to potential security issues, I think we should not allow relative path and sandbox to the agent directory.

**Comment by urso** [comment]
yes, potential security issues by itself. Especially ragarding the log collection setup. But also validating and rejecting inputs if we figure they have a special pattern. For example the worst would be a template that puts a user variable before the type. If we do not protect from string injection, then I could create a string that produces whatever configuration I want using a label on my container (send me your secrets).

**Comment by ph** [comment]
Yes, but I am not sure how you could prevent that, what mechanism we could put in place to reduce this risk?

**Comment by urso** [comment]
Templates proposed so far:

```
inputs:
  - type: logfile
    path: ${template.paths.syslog|'/var/log/syslog'}
```

```
inputs:
  - type: redis/metrics
    streams:
    - data_stream:
        dataset: redis.info
        type: metrics
      metricsets:
        - info
      hosts:
        - ${template.host}:${template.port}  # we can only get 1 host?
      username: '${template.username}'
      period: '${template.period}'
      condition: ${template.metricset.info} == 'true'
   - data_stream:
        dataset: ...
```

Judging by the format we do parse the configuration into an 'AST' on init and apply the variables to the abstract AST, which would be an improvement over plain text based templates when it comes to potential string injection attacks. Still one might be able to pass insecure settings, especially if the Agent is run as root. Here the template author might want to restrict the settings even further (e.g. URL paths...). It seems with dynamic input like configurations, template authors have no support for:
- Loops?
- Selectively add/remove settings
- restrict settings based on other settings (or hardcoded conditions).
- Add custom input validation
- Add custom error reporting
- Combine/Add/Remove array based settings in a customized order (processor settings)

Alternatively to plain text templates or our own restricted evaluation similar to dynamic inputs, we might consider to empower template authors by providing them a (restricted) scripting language that has been developed for configuration management/templating. For example `lua` (our deps already contain different lua dialects) or [`startlark`](https://github.com/google/starlark-go) (I'm leaning towards the later).

**Comment by ChrsMark** [comment]
+1 for having an option to restrict access to URLs. Specially when using `http` input. For example in k8s we want to restrict users from being able to access k8s API using an `http` input through hints. We can introduce `allow`/`deny` lists for host setting on kubernetes provider and and cross check them with what users provide using hints so as to block them from accessing sensitive endpoints like k8s API. 

ie:
```
provider.kuberentes:
   denylist.host: [10.250.13.7] # 10.250.13.7 is the k8s API address
```

This could work on k8s and with specific examples that we have seen so far, however having something more generic (not k8s related) would be useful too.

cc: @exekias @jsoriano

**Comment by urso** [comment]
Reason for generic functionality is that we might want to apply even more custom limits for generic inputs like http or logs input. e.g. the http input has support to read a bearer token from disk. An admin or template author should have support to restrict the path for the bearer token using a template.

**Comment by exekias** [comment]
++, I could also see how we may see the need for more logic in templates in the future.

For templates it sounds like some kind of type + semantic validation for input parameters would be enough to achieve some level of security.

To some extent, this allow/deny feature sounds like something you want in general, right? Also outside of the scope of templates. Would it make sense to think about it in the context of agent & fleet? As a user I may also want to limit the configs that can come from Fleet itself. We could be using the same feature here

**Comment by urso** [comment]
Allow/denylists totally make sense at the input/integration level. With templates I would expect template authors to wish to be more strict or even introduce some additional configurable checks.

**Comment by mukeshelastic** [comment]
> yes, potential security issues by itself. Especially ragarding the log collection setup. But also validating and rejecting inputs if we figure they have a special pattern. For example the worst would be a template that puts a user variable before the type. If we do not protect from string injection, then I could create a string that produces whatever configuration I want using a label on my container (send me your secrets).

This is the disadvantage of templates, it enables edge to pass variables for malicious exploits. @urso thanks for raising and looking deeper into this.

From priority perspective, I'd like to separate  the core standalone and K8s autodiscover functionality from security. For first release, we should proceed with templates that allow hints variables passthrough to enable hints based auto discovery. Parallely we should investigate the security problems with the approach and what we should do to prevent malicious usage. We can decide to keep the input templates as beta until we feel comfortable with the security risk.

**Comment by ruflin** [comment]
++ on the comment from @mukeshelastic . The initial goal is to have a replacement for what we have today in Beats. AFAIK all the above restrictions don't exist in Beats today.

**Comment by exekias** [comment]
That's correct, if you think about it, this is only used by people with access to deploying in the cluster (k8s) or configuring the agent (standalone), we don't put high security requirements there.

For the case of kubernetes this feature is disabled by default, just in case, as only some admins will want to enable their users to spawn inputs on their own.
